libvast/vast/concept/parseable/vast/expression.hpp

```diff
@@ -165,23 +165,22 @@ struct expression_parser : parser<expression_parser> {
       | "&&"_p  ->* [] { return logical_and; }
       ;
     expr
-      = ((group >> *(ws >> and_or >> ws >> group)) ->* to_expr)
-      >> ws >> parsers::eoi
+      = (group >> *(ws >> and_or >> ws >> group) >> ws) ->* to_expr
       ;
     // clang-format on
     return expr;
   }
 
-  template <class Iterator>
-  bool parse(Iterator& f, const Iterator& l, unused_type) const {
+  template <class Iterator, class Attribute>
+  bool parse(Iterator& f, const Iterator& l, Attribute& x) const {
     static auto p = make<Iterator>();
-    return p(f, l, unused);
-  }
-
-  template <class Iterator>
-  bool parse(Iterator& f, const Iterator& l, expression& a) const {
-    static auto p = make<Iterator>();
-    return p(f, l, a);
+    if (!p(f, l, x))
+      return false;
+    if (f != l) {
+      VAST_ERROR_ANON("failed to parse expression", std::string(f, l));
+      return false;
+    }
+    return true;
```

Comment:
```suggestion
    return f == l;
```

We usually don't have side effects like logging a parser. This should go into the parent scope.

The error message says `expression` but the signature has an unconstrained template parameter. Is this intentional?

Reply:
> The error message says `expression` but the signature has an unconstrained template parameter. Is this intentional?

We do this for other parsers as well, since the input may be either `expression` or `unused_type`.

> We usually don't have side effects like logging a parser. This should go into the parent scope.

It can't go into the parent scope. This log message shows _where_ the parser failed when parsing an expression. I know this is suboptimal, but getting a generic "parser_error !!" message when the expression parser fails isn't optimal either. I think it's a reasonable tradeoff.

We do have a task somewhere to allow carrying an error message (or at least an errc à la `std::from_chars`) through our parser framework, but that is a larger refactoring.